File: go/300.500/451.SortCharactersByFrequency.go
package main

import (
	"bytes"
	"fmt"
	"sort"
)

func frequencySort(s string) string {
	mp := map[byte]int{}
	for i := range s {
		mp[s[i]]++
	}
	type pair struct {
		c   byte
		cnt int
	}
	pairs := make([]*pair, 0, len(mp))
	for k, v := range mp {
		pairs = append(pairs, &pair{k, v})
	}
	sort.Slice(pairs, func(i, j int) bool {
		return pairs[i].cnt > pairs[j].cnt
	})
	ans := make([]byte, 0, len(s))
	for _, p := range pairs {
		ans = append(ans, bytes.Repeat([]byte{p.c}, p.cnt)...)
	}
	return string(ans)
}

func frequencySort1(s string) string {
	mp := map[byte]int{}
	maxf := 0
	for i := range s {
		mp[s[i]]++
		maxf = max(maxf, mp[s[i]])
	}
	buckets := make([][]byte, maxf+1)
	for c, cnt := range mp {
		buckets[cnt] = append(buckets[cnt], c)
	}
	ans := make([]byte, 0, len(s))
	for i := maxf; i > 0; i-- {
		for _, ch := range buckets[i] {
			ans = append(ans, bytes.Repeat([]byte{ch}, i)...)
		}
	}
	return string(ans)
}

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}

func main() {
	fmt.Println("tree", frequencySort("tree"))
	fmt.Println("cccaaa", frequencySort("cccaaa"))
	fmt.Println("Aabb", frequencySort("Aabb"))

	fmt.Println("tree", frequencySort1("tree"))
	fmt.Println("cccaaa", frequencySort1("cccaaa"))
	fmt.Println("Aabb", frequencySort1("Aabb"))
}

File: 300.500/451.SortCharactersByFrequency.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <algorithm>
#include <vector>
using namespace std;

class SortCharactersByFrequency {
public:
    string frequencySort(string s) {
        unordered_map<char, int> mp;
        for(char c : s){
            mp[c]++;
        }
        vector<pair<char, int>> vec;
        for(auto &kv : mp){
            vec.emplace_back(kv);
        }
        sort(vec.begin(), vec.end(), [](const pair<char, int> &a,const pair<char, int> &b){
            return a.second > b.second;
        });
        string ret = "";
        for(auto &p : vec){
            for(int i =0;i< p.second;++i){
                ret.push_back(p.first);
            }
        }
        return ret;
    }
public:
    string frequencySort1(string s) {
        int maxf = 0;
        unordered_map<char, int> mp;
        for(char c : s){
            maxf = max(maxf, ++mp[c]);
        }
        //注意这里 相同出现次数的字符 可能会有多个
        vector<string> bucket(maxf+1, "");
        for(auto &kv : mp){
            bucket[kv.second].push_back(kv.first);
        }
        string ret = "";
        for(int i= maxf; i > 0;--i){
            string &b = bucket[i];
            for(char c : b){
                for(int j =0;j<i;++j){
                    ret.push_back(c);
                }
            }
        }
        return ret;
    }
};

int main(){
    SortCharactersByFrequency cls;
    cout << "tree -> " << cls.frequencySort("tree").c_str() << endl;
    cout << "cccaaa -> " << cls.frequencySort("cccaaa").c_str() << endl;
    cout << "Aabb -> " << cls.frequencySort("Aabb").c_str() << endl;

    cout << "tree -> " << cls.frequencySort1("tree").c_str() << endl;
    cout << "cccaaa -> " << cls.frequencySort1("cccaaa").c_str() << endl;
    cout << "Aabb -> " << cls.frequencySort1("Aabb").c_str() << endl;
    return 0;
}